test/src/tests-session-manager.ts

```diff
@@ -0,0 +1,607 @@
+            const persistenceSpy = sinon.spy(mpInstance._Persistence, 'update');
+
+            mpInstance._SessionManager.endSession();
+
+            expect(mpInstance._Store.sessionId).to.equal(null);
+            expect(mpInstance._Store.dateLastEventSent).to.equal(null);
+            expect(mpInstance._Store.sessionAttributes).to.eql({});
+
+            // Persistence isn't necessary for this feature, but we should test
+            // to see that it is called in case this ever needs to be refactored
+            expect(persistenceSpy.called).to.equal(true);
+        });
+
+        it('should force a session end when override is used', () => {
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+            const persistenceSpy = sinon.spy(mpInstance._Persistence, 'update');
+
+            mpInstance._SessionManager.endSession(true);
+
+            expect(mpInstance._Store.sessionId).to.equal(null);
+            expect(mpInstance._Store.dateLastEventSent).to.equal(null);
+            expect(mpInstance._Store.sessionAttributes).to.eql({});
+
+            // Persistence isn't necessary for this feature, but we should test
+            // to see that it is called in case this ever needs to be refactored
+            expect(persistenceSpy.called).to.equal(true);
+        });
+
+        it('should do nothing when a session does not exist', () => {});
+        it('should do nothing when persistence cookies are missing', () => {});
+        it('should nullify session attributes when override is used', () => {});
+        it('should use the cookie sessionId over Store.sessionId', () => {});
+
+        it('should NOT end session if session has not timed out', () => {
+            const now = new Date();
+            const twentyMinutesAgo = new Date();
+            twentyMinutesAgo.setMinutes(now.getMinutes() - 20);
+
+            mParticle.init(apiKey, window.mParticle.config);
+            const mpInstance = mParticle.getInstance();
+            const timerSpy = sinon.spy(
+                mpInstance._SessionManager,
+                'setSessionTimer'
+            );
+
+            // Session Manager relies on persistence to determine last time seen (LES)
+            // Also requires sid to verify session exists
+            sinon.stub(mpInstance._Persistence, 'getPersistence').returns({
+                gs: {
+                    les: twentyMinutesAgo,
+                    sid: 'fake-session-id',
+                },
+            });
+
+            mpInstance._SessionManager.endSession();
+
+            expect(mpInstance._Store.sessionId).to.equal('fake-session-id');
+
+            // When session is not timed out, setSessionTimer is called to keep track
+            // of current session timeout
+            expect(timerSpy.getCalls().length).to.equal(1);
+        });
+
+        it('should end session if session times out', () => {
```

Comment:
I see what you are doing here, but I think the approach is a bit off.  By calling `endSession` you are not letting the session time out on its own, but instead forcing a session to end.

Instead, you should initialize mParticle, check to see if `sessionId`, `dateLastEventSet` exist, and then move the clock forward (`clock.tick(500000)`) or whatever number it is that exceeds the default session timeout. Then you can call the expects you have on lines 336-342 again.

You can keep stubbing persistence as needed though in line 327.